jax/core.py

```diff
@@ -1207,16 +1207,165 @@ def raise_to_shaped(aval: AbstractValue, weak_type=None):
+    """
+    return d1 == d2
+
+  def greater_equal(self, d1: DimSize, d2: DimSize) -> bool:
+    """Computes `d1 >= d2`.
+    Raise InconclusiveDimensionOperation if the result is different in
+    different contexts.
+    """
+    return d1 >= d2
+
+  def sum(self, *ds: DimSize) -> DimSize:
+    """Sum of dimensions.
+    Raises InconclusiveDimensionOperation if the result cannot be represented
+    by the same DimSize in all contexts.
+    """
+    return sum(ds)
+
+  def diff(self, d1: DimSize, d2: DimSize) -> DimSize:
+    """Difference of dimensions.
+    Raises InconclusiveDimensionOperation if the result cannot be represented
+    by the same DimSize in all contexts.
+    """
+    return d1 - d2
+
+  def divide_shape_sizes(self, s1: Shape, s2: Shape) -> int:
+    """Computes the division of the sizes of the shapes.
+
+    Raise InconclusiveDimensionOperation if the result is different in different
+    contexts, or if the division is not even.
+    """
+    sz1 = np.prod(s1)
+    sz2 = np.prod(s2)
+    if sz1 == 0 and sz2 == 0:
+      return 1
+    if sz1 % sz2:
+      raise InconclusiveDimensionOperation(f"Cannot divide evenly the sizes of shapes {tuple(s1)} and {tuple(s2)}")
+    return sz1 // sz2
+
+  def stride(self, d: DimSize, window_size: DimSize, window_stride: DimSize) -> DimSize:
+    """(d - window_size) // window_stride + 1"""
+    return (d - window_size) // window_stride + 1
+
+  def dilate(self, d: DimSize, dilation: int) -> DimSize:
+    """Implements `0 if d == 0 else 1 + dilation * (d - 1))`"""
+    return 0 if d == 0 else 1 + dilation * (d - 1)
+
+
+_dimension_handler_int = DimensionHandler()
+_SPECIAL_DIMENSION_HANDLERS: Dict[type, DimensionHandler] = {}
+
+def _get_dim_handler(*dlist: DimSize) -> DimensionHandler:
+  """Finds the handler that works for all dimension sizes.
+  At most one special dimension type is allowed. Non-special dimension
+  must be convertible to integers.
+  """
+  special_handlers = set()
+  for d in dlist:
+    handler = _SPECIAL_DIMENSION_HANDLERS.get(type(d))
+    if handler:
+      special_handlers.add(handler)
+    else:
+      # We check that the dimension is convertible to an int
+      # TODO: better error message
+      operator.index(d)
+  if special_handlers:
+    handler, *others = special_handlers
+    if others:
+      msg = (f"Dimension size operation involves multiple non-int types {dlist}")
+      raise TypeError(msg)
+    return handler
+  else:
+    return _dimension_handler_int
+
+def symbolic_equal_dim(d1: DimSize, d2: DimSize) -> bool:
+  return _get_dim_handler(d1, d2).symbolic_equal(d1, d2)
+
+def symbolic_equal_one_of_dim(d1: DimSize, dlist: Sequence[DimSize]) -> bool:
+  handler = _get_dim_handler(d1, *dlist)
+  return any([handler.symbolic_equal(d1, d2) for d2 in dlist])
+
+def symbolic_equal_shape(s1: Shape, s2: Shape) -> bool:
+  """See DimensionHandler.symbolic_equal."""
+  return (len(s1) == len(s2) and
+          all(safe_map(symbolic_equal_dim, s1, s2)))
```

Comment:
IMO it's better to write this as just `map`, since `safe_map` should be the default (and `unsafe_map`, aka builtin `map`, should be opt-in), but it's not important.

Reply:
Done